### ISSUE TITLE
Interpolated value in join not accessable as parameters in stored procedure

### ISSUE THREAD

**Issue by madshargreave** [elixir-ecto/ecto#1791]
### Environment

Elixir 1.3.2
Postgres 9.5
Ecto. 2.0.5
Postgrex
OSX

### Current behavior

I have the following snippet of a query
```

 def find_by_conversation_with_groups(query \\ Visitor, user_id, conversation_id) do
    query =
      query
      |> join(:left, [v], c in assoc(v, :conversations))
      |> join(:left, [v, c], a in assoc(c, :app))
      |> join(:left, [v, c, a], u in assoc(a, :users))
      |> join(:left, [v, c, a, u], g in assoc(u, :groups), g.user_id == ^user_id and g.user_id == u.id)
      |> join(:inner_lateral, [v, c, a, u, g],
        matches in fragment("SELECT * from find_matching_visitor(?, ?, ?)", v.id, g.formatted_default_filters, g.formatted_custom_filters),
        matches.id == v.id)

    from [v, c, a, u, g] in query,
    where: c.id == ^conversation_id and not ilike(g.name, "all"),
    group_by: [v.id, c.id],
    select: %{
      id: v.id,
      app_id: v.app_id,
     ...
```

When executed, it will cast the following error

```
** (ArgumentError) Postgrex expected a binary that can be encoded/cast to type "text", got 100139. Please make sure the value you are passing matches the definition in your table or in your query or convert the value accordingly.
```

Then, by inspecting the generated query, it appears the the `user_id` argument that is interpolated in the `groups` join, is not carried over into the parameters to the SQL function as can be seen below:

```
[debug] QUERY ERROR source="visitors" db=15.9ms
SELECT v0."id", v0."app_id", max(c1."id"), v0."first_seen", v0."last_seen", v0."signed_up_at", v0."referrer", v0."last_heard_from", v0."last_contacted", v0."last_page_viewed", v0."sessions", v0."seen_welcome_message", v0."email", v0."location", v0."name", v0."slug", v0."type", v0."avatar", array_remove(array_agg(g4."name"), NULL) FROM "visitors" AS v0 LEFT OUTER JOIN "conversations" AS c1 ON c1."visitor_id" = v0."id" LEFT OUTER JOIN "apps" AS a2 ON a2."id" = c1."app_id" LEFT OUTER JOIN "app_users" AS a6 ON a6."app_id" = a2."id" LEFT OUTER JOIN "users" AS u3 ON a6."user_id" = u3."id" LEFT OUTER JOIN "groups" AS g4 ON (g4."user_id" = u3."id") AND ((g4."user_id" = $1) AND (g4."user_id" = u3."id")) INNER JOIN LATERAL (SELECT * from find_matching_visitor(v0."id", g4."formatted_default_filters", g4."formatted_custom_filters")) AS f5 ON f5."id" = v0."id" WHERE ((c1."id" = $2) AND NOT (g4."name" ILIKE 'all')) GROUP BY v0."id", c1."id" [100139]
```

**Comment by josevalim** [comment]
What is the type of the user_id columns for the Group schema and in the database in the "groups" table?

**Comment by madshargreave** [comment]
String

**Comment by josevalim** [comment]
Can you please include the schema for Group? It seems the `user_id` value is an integer, can you confirm that? That wouldn't match your reply of it being aa UUID.

In any case, such mismatches was supposed to be caught by the query planner.

**Comment by madshargreave** [comment]
Group schema:

```
schema "groups" do
    field :name, :string
    field :filters, {:array, :map}, on_replace: :delete
    field :formatted_default_filters, {:array, :string}, on_replace: :delete
    field :formatted_custom_filters, {:array, :string}, on_replace: :delete
    field :default, :boolean

    belongs_to :user, User, type: :string
    belongs_to :app, App, type: :string

    timestamps
  end
```

**Comment by josevalim** [comment]
Alright, so this doesn't make sense. This was supposed to error before reaching Postgrex. Do you have an easy way to reproduce this so we can further investigate it? Alternatively you can quickly try Ecto 2.1 and see if it has been fixed.

**Comment by madshargreave** [comment]
It appears to have been solved in Ecto 2.1. Is this safe for production-usage at this stage?

**Comment by josevalim** [comment]
Awesome.

Yes, it is. The only reason it has not been released yet is because we are waiting on Postgrex to become 1.0. It should be out this or next month.
